### PR TITLE
create routeTable and routeManager API

### PR DESCRIPTION
**Pull request by walterddr**
currently there's no way to get a routing table out without creating a dummy table broker request. 

- hoist the routing table and routing manager into  pinot-core as interface.
- creating a new API to get routing table.

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java [hunk 1]
```diff
@@ -458,13 +460,42 @@ public boolean routingExists(String tableNameWithType) {
...
+    if (routingEntry == null) {
+      return null;
+    }
+    InstanceSelector.SelectionResult selectionResult = routingEntry.calculateRouting(tableNameWithType);
+    return constructRoutingTableFromInstanceSelectionResult(tableNameWithType, selectionResult);
+  }
+
+  @Override
+  public Map<String, ServerInstance> getEnabledServerInstanceMap() {
+    return new HashMap<>(_enabledServerInstanceMap);
```
Comment (siddharthteotia) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java:486]: Can we add a comment to explain why copy for readers (or in future whoever changes the code) to be aware of the gotchas ?

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  RouteTable getRoutingTable(String tableNameWithType);
+
+  /**
+   * Get the {@link RouteTable} for a specific broker request.
+   *
+   * @param brokerRequest the broker request constructed from a query.
+   * @return the route table.
+   */
+  RouteTable getRoutingTable(BrokerRequest brokerRequest);
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java:64]: Can we just keep this interface ? Why do we need the above table level API ?

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java (hunk 1 above)
Comment (walterddr) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java:486]: +1. yeah.

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java (hunk 2 above)
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java:64]: not really. otherwise one will have to basically create a "SELECT * FROM <tableName>` dummy query, generate a brokerRequest, and then gets the routing table. then inside the getRouting table method it needs to parse the brokerrequest. parse out all the filter clauses (which is none) then apply pruning logic which is again unnecessary, then return back the full routing table

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java [hunk 3]
```diff
@@ -82,7 +83,7 @@
  * TODO: Expose RoutingEntry class to get a consistent view in the broker request handler and save the redundant map
  *       lookups.
  */
-public class RoutingManager implements ClusterChangeHandler {
+public class RoutingManager implements RouteManager, ClusterChangeHandler {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java:86]: Let's name the interface `RoutingManager` and rename the implementation to `DefaultRoutingManager`

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingTable.java [hunk 4]
```diff
@@ -20,10 +20,11 @@
 
 import java.util.List;
 import java.util.Map;
+import org.apache.pinot.core.routing.RouteTable;
 import org.apache.pinot.core.transport.ServerInstance;
 
 
-public class RoutingTable {
+public class RoutingTable implements RouteTable {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingTable.java:27]: Name the interface `RoutingTable` and rename the implementation to `RoutingTableImpl`

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelector.java [hunk 5]
```diff
@@ -56,7 +55,7 @@
    * to selected server instance hosting the segment and a set of unavailable segments (no enabled instance or all
    * enabled instances are in ERROR state).
    */
-  SelectionResult select(BrokerRequest brokerRequest, List<String> segments);
+  SelectionResult select(List<String> segments);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelector.java:58]: Don't change this API. The future implementation of this might need the query to correctly select the instance

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java:486]: Don't make a copy as it can be expensive, just return as is since we can control the caller code

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+   */
+  Map<String, ServerInstance> getEnabledServerInstanceMap();
+
+  /**
+   * Get the {@link RouteTable} for a specified table.
+   *
+   * @param tableNameWithType the name of the table.
+   * @return the route table.
+   */
+  RouteTable getRoutingTable(String tableNameWithType);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java:56]: We can provide a default implementation to compile `SELECT * FROM <tableNameWithType>` and call `getRoutingTable()` with `brokerRequest`

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java (hunk 1 above)
Comment (walterddr) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/RoutingManager.java:486]: we have to make a copy since if one is looping over the entry during which the content of the map is changed due to helix update. we will not be able to guarantee correctness.

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java (hunk 6 above)
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java:56]: yeah I think i replied above. this basically 
- makes the code to create a SELECT * FROM <tableName> dummy query, generate a brokerRequest, and then gets the routing table. 
- then inside the getRouting table method it needs to parse the brokerrequest. parse out all the filter clauses (which is none) then apply pruning logic which is again unnecessary, then return back the full routing table

These are all unnecessary overheads to maintain correctness.

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java:56]: I think this is because we are thinking of a simple JOIN query where the JOIN has to be performed by considering all segments of both tables so may be that's why we are thinking of a SELECT * type query in which case a dummy / default implementation of `getRoutingTable`() using `tableNameWithType` makes sene

However, consider a colocated join for two tables partitioned on the join key, the partition pruning has to happen / pushed below the join so the segment selector followed by instance selector are aware of the exact pruned segments that are going to participate in JOIN. 

In that case, we need the existing generic `BrokerRequest` based `getRoutingTable()` API imo.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/BrokerRoutingManager.java [hunk 7]
```diff
@@ -458,13 +461,32 @@ public boolean routingExists(String tableNameWithType) {
...
+  private String getIdealStatePath(String tableNameWithType) {
+    return _idealStatePathPrefix + tableNameWithType;
+  }
+
+  private String getExternalViewPath(String tableNameWithType) {
+    return _externalViewPathPrefix + tableNameWithType;
+  }
+
+  private RoutingTable constructRoutingTableFromInstanceSelectionResult(String tableNameWithType,
+      InstanceSelector.SelectionResult selectionResult) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/BrokerRoutingManager.java:489]: Shall we revert this part of change as they are not related?

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/BrokerRoutingManager.java [hunk 8]
```diff
@@ -458,13 +461,32 @@ public boolean routingExists(String tableNameWithType) {
...
     if (routingEntry == null) {
       return null;
     }
     InstanceSelector.SelectionResult selectionResult = routingEntry.calculateRouting(brokerRequest);
+    return constructRoutingTableFromInstanceSelectionResult(tableNameWithType, selectionResult);
+  }
+
+  @Override
+  public Map<String, ServerInstance> getEnabledServerInstanceMap() {
+    return new HashMap<>(_enabledServerInstanceMap);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/BrokerRoutingManager.java:477]: Directly return the `_enabledServerInstanceMap` to avoid the unnecessary overhead

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentselector/SegmentSelector.java [hunk 9]
```diff
@@ -61,4 +61,9 @@
    * (table) without overlap.
    */
   Set<String> select(BrokerRequest brokerRequest);
+
+  /**
+   * Selects the segments based on the table name.
+   */
+  Set<String> select(String tableName);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentselector/SegmentSelector.java:68]: Revert this and all the changes in the implementation

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java [hunk 10]
```diff
@@ -263,7 +262,7 @@ void updateSegmentMaps(IdealState idealState, ExternalView externalView, Set<Str
   }
 
   @Override
-  public SelectionResult select(BrokerRequest brokerRequest, List<String> segments) {
+  public SelectionResult select(List<String> segments) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java:265]: Revert this

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RoutingManager.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+   * @return true if the route table exists.
+   */
+  boolean routingExists(String tableNameWithType);
+
+  /**
+   * Get all enabled server instances that are available for routing.
+   *
+   * @return all currently enabled server instances.
+   */
+  Map<String, ServerInstance> getEnabledServerInstanceMap();
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/routing/RoutingManager.java:48]: (minor) Move this API to the front because `routingExists` and `getRoutingTable` are related and better to be put together for readability

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RoutingManager.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+/**
+ * The {@code RouteManager} provides the routing information for a query that requests access to a Pinot table.
+ *
+ * The implementation of this interface should ensure the routing and server information are up-to-date at the
+ * time when the routing request was made.
+ *
+ * set by the user. This needs to be added to support features like segment pruning.
+ */
+public interface RoutingManager {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/routing/RoutingManager.java:33]: Annotate it with:
@InterfaceAudience.Public
@InterfaceStability.Evolving

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java (hunk 10 above)
Comment (walterddr) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java:265]:
```suggestion
  public SelectionResult select(BrokerRequest brokerRequest, List<String> segments) {
```

File: pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java (hunk 6 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/routing/RouteManager.java]: @walterddr  looks like this concern will be addressed based on the latest discussion that a `StageNode` based `getRoutingTable` API can be added in the future ?

**Review by siddharthteotia — APPROVED** [review]
